### ISSUE TITLE
Read-only List type

### ISSUE THREAD

**Issue by astigsen** [realm/realm-cocoa#4985]
It would be really useful to have a `List` type that is read-only.

Sometimes your Model contains properties that should only be modified by other processes or with sync, only by the server. It would be really nice to be able to specify that in the type system, so that can't accidentally end up writing to them.

With regular properties you can do that by specifying them as `dynamic private(set) var myproperty = ""`, but that does not work with `List`.

**Comment by jpsim** [comment]
This would certainly be useful, but I'd rather consider the problem in the general case before coming up with individualized solutions, like this case for `List`.

For example, even with the private setter approach you mention above (which is also possible to do in Objective-C), Realm still doesn't know if the property should be settable via other means, such as `setValue(_:forKey:)`.

We should design a system in which we don't cause an explosion of types (mutable & immutable variants for each property type).

**Comment by astigsen** [comment]
Yes, do you have any ideas for how this could be implemented? We could obviously do our own totally separate annotation, like we do for ignored properties. But since that does not involve the type system, it will only give runtime errors.

**Comment by jpsim** [comment]
I have some ideas that would require making `List` & `RealmOptional` a `struct` (still backed by a `class`), in which mutating operations would be marked as `mutating`, offering the compile-time validation you're looking for.

Another (longer term) idea would rely on Swift's work-in-progress property behaviors to add compile-time mutability annotations to Realm types.
